### PR TITLE
DreamOS

### PR DESCRIPTION
**Pull request by dinosaure**
This is a simple example of a DreamOS which has several limitation:
- [ ] it handles only HTTP/1.1 but the support of H2 will be easy
- [ ] it does not support upgrade connection to `websocket` (it's a bit too divergent from what I did in `paf`)
- [x] it includes `cohttp` for `letsencrypt` (but I prepared something to completely delete `cohttp` from it here: mmaker/ocaml-letsencrypt#26)
- [ ] `paf` is not released yet but close to be
- [ ] complete the interface of `dream.mirage` to be like `dream` (without `static`)
- [ ] make a `static` middleware which take a `(module Mirage_kv.S)` implementation
- [ ] `*.eml` is not handled but we can I think with `ocamlbuild` (or waiting MirageOS 4)

The current example is deployed here - with `gcloud` (with a non-productive certificate!): https://dream.egar.im/
I will have less and less time in the coming weeks but it's a nice start for DreamOS 🎉 ! If someone would like to participate (mostly to copy/paste the H2 handler), it will be really appreciate.

### CODE REVIEW

**Comment by aantron** [comment]
Looks nice! The overall structuer of the PR seems quite reasonable. I will give it a detailed look over the next few days and leave comments.

**Comment by aantron** [comment]
@dinosaure For most of this PR, I'm ready to merge and have you "take over" `src/mirage` and Mirage examples, so you can do what is necessary for the port :) The code looks mostly self-contained. I can raise various questions, but I don't think blocking the PR is necessary.

The current issue is: if I run `dune build -p dream` in this branch, I get

```
File "src/vendor/dune", line 8, characters 36-41:
8 |   (libraries faraday bigstringaf ke mimic)))
                                        ^^^^^
Error: Library "mimic" not found.
Hint: try:
  dune external-lib-deps --missing -p dream @install
File "src/vendor/dune", line 22, characters 46-58:
22 |   (libraries dream.paf tls-mirage mirage-time mirage-stack dream.paf.alpn)))
                                                   ^^^^^^^^^^^^
Error: Library "mirage-stack" not found.
Hint: try:
  dune external-lib-deps --missing -p dream @install
File "src/vendor/dune", line 29, characters 49-59:
29 |   (libraries ipaddr domain-name dream.paf httpaf cohttp-lwt)))
                                                      ^^^^^^^^^^
Error: Library "cohttp-lwt" not found.
Hint: try:
  dune external-lib-deps --missing -p dream @install
File "src/vendor/dune", line 38, characters 4-15:
38 |     letsencrypt)))
         ^^^^^^^^^^^
Error: Library "letsencrypt" not found.
Hint: try:
  dune external-lib-deps --missing -p dream @install
```

So some Mirage packages have "infected" the basic build.

**Comment by aantron** [comment]
Let me know if you want commentary on the details of the Mirage version of the API, etc. I am currently assuming we will just discuss those in future work.

**Comment by dinosaure** [comment]
> So some Mirage packages have "infected" the basic build.

Yes, I currently wants to delete `cohttp` as an "insidious" dependency from `letsencrypt`. I did not upgrade the Dream's OPAM file and it's why the build fail. My goal is to:
- release `letsencrypt` without `cohttp`
- release `paf` with the new version of `letsencrypt` (and split `paf.cohttp` to `paf-cohttp`)
- upgrade the Dream's OPAM file with all of that

**Comment by aantron** [comment]
AFAIK `paf` is currently only for the Mirage version of Dream. If that's correct, then `dream` shouldn't depend on it. Do you intend to use package name `dream` for the Mirage version?

**Comment by dinosaure** [comment]
> AFAIK paf is currently only for the Mirage version of Dream. If that's correct, then dream shouldn't depend on it. Do you intend to use package name dream for the Mirage version?

Currently, the MirageOS support exists via `dream.mirage`. So this PR brings new dependencies needed for `dream.mirage` from the OPAM point-of-view. I need to do some tries to see if a split still work or not - I remember some issues when I tried to make a new `dream-mirage.opam` package (mainly about access of internal stuffs).

**Comment by aantron** [comment]
> I remember some issues when I tried to make a new `dream-mirage.opam` package (mainly about access of internal stuffs).

What were the issues? I just tried creating `dream-mirage.opam` and moving `dream.mirage` to `dream-mirage`. The issues I got were unchanged from not having the split (`src/vendor/dune` appears to depend on `paf`, etc.)

Could you push your nearest attempt into this PR? If you can separate the vendoring, I can probably take care of all the internal references issues.

**Comment by dinosaure** [comment]
I will do a try soon, currently we are are able to delete `cohttp` from the dependency graph with the new release of `letsencrypt`.

**Comment by dinosaure** [comment]
I just created `dream-mirage` package and it's fine :+1:. Then, I currently updated the whole stack to use `letsencrypt.0.3.0` and delete by this way the CoHTTP dependency and the unikernel become smaller: from 8.6Mo to 7.9Mo 🎉 . The `h2` supports still is needed but, as I said, it should be easy to implement it.

**Comment by aantron** [comment]
Based on this [error in CI](https://github.com/aantron/dream/pull/119/checks?check_run_id=3125029215#step:7:12):

```
File "src/mirage/dune", line 2, characters 14-26:
2 |  (public_name dream-mirage)
                  ^^^^^^^^^^^^
Error: The current scope doesn't define package "dream-mirage".
The only packages for which you can declare elements to be installed in this
directory are:
- dream (because of dream.opam)
```

You may need to add `dream-mirage.opam` or the like. You may already have seen that :)

File: src/mirage/mirage.mli [hunk 1]
```diff
@@ -84,23 +96,99 @@ module Make
...
 
   val get : string -> handler -> route
   val not_found : handler
 
+  type error =
+    { condition :
+        [ `Response of response
+        | `String of string
+        | `Exn of exn ]
+    ; layer :
```
Comment (aantron) [inline src/mirage/mirage.mli:164]: Not a blocker for merging, but: can you use the `.ocamlformat` from this repo? Or is this formatted manually? The `;`-first style AFAIK dates from when OCaml didn't have trailing separators, and mimics languages like Elm which don't have them. It's very unlike JS/JSON style, is not Web-friendly.  If all of that is true, this style is bakcwards. I avoid it in Dream. See also the [Dream docs](https://aantron.github.io/dream/#type-error).

Compare with the definition of this type:

https://github.com/aantron/dream/blob/62d7cd48327bbc1d7ee5792c2d85547b995e31f9/src/pure/inmost.ml#L89-L118

in particular, note the final `;` on the last field

https://github.com/aantron/dream/blob/62d7cd48327bbc1d7ee5792c2d85547b995e31f9/src/pure/inmost.ml#L117-L118

which makes the whole `;`-first style pointless. The version in this PR is inconsistent with Dream, Web programming, and the OCamlFormat output on this type using this repo's config:

```ocaml
type error = {
  condition : [`Response of response | `String of string | `Exn of exn];
  layer : [`App | `HTTP | `HTTP2 | `TLS | `WebSocket];
  caused_by : [`Server | `Client];
  request : request option;
  response : response option;
  client : string option;
  severity : log_level;
  debug : bool;
  will_send_response : bool;
}
```

File: src/mirage/mirage.ml [hunk 2]
```diff
@@ -1,3 +1,5 @@
+[@@@warning "-32"]
```
Comment (aantron) [inline src/mirage/mirage.ml:1]: What exactly is this suppressing? Can it be made more local?

File: dream-mirage.opam [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+bug-reports: "https://github.com/aantron/dream/issues"
+dev-repo: "git+https://github.com/aantron/dream.git"
+
+author: "Anton Bachin <antonbachin@yahoo.com>"
+maintainer: "Anton Bachin <antonbachin@yahoo.com>"
+
+depends: [
+  "dream"
+  "dune" {>= "2.7.0"}
+  "dream"
```
Comment (aantron) [inline dream-mirage.opam:52]: `dream` is listed twice. Not a blocker.

**Review by aantron — COMMENTED** [review]
Thanks! I left some comments which are more like questions/notes that don't block this PR, but can be addressed in the future or at least discussed.

File: src/mirage/mirage.mli (hunk 1 above)
Comment (dinosaure) [reply src/mirage/mirage.mli:164]: I don't have strong opinion about style :+1:.

File: src/mirage/mirage.ml (hunk 2 above)
Comment (dinosaure) [reply src/mirage/mirage.ml:1]: I need to lint a bit the whole API to see what is really useless and what is really needed but not yet used.

**Pull request merged**